mono/metadata/appdomain.c

```diff
@@ -706,56 +706,102 @@ mono_domain_has_type_resolve (MonoDomain *domain)
 
 /**
  * mono_domain_try_type_resolve:
- * \param domain application domainwhere the name where the type is going to be resolved
+ * \param domain application domain in which to resolve the type
  * \param name the name of the type to resolve or NULL.
  * \param tb A \c System.Reflection.Emit.TypeBuilder, used if name is NULL.
  *
  * This routine invokes the internal \c System.AppDomain.DoTypeResolve and returns
- * the assembly that matches name.
- *
- * If \p name is null, the value of \c ((TypeBuilder)tb).FullName is used instead
+ * the assembly that matches name, or ((TypeBuilder)tb).FullName.
  *
  * \returns A \c MonoReflectionAssembly or NULL if not found
  */
 MonoReflectionAssembly *
 mono_domain_try_type_resolve (MonoDomain *domain, char *name, MonoObject *tb)
 {
 	MonoError error;
-	MonoReflectionAssembly *ret = mono_domain_try_type_resolve_checked (domain, name, tb, &error);
+	g_assert (domain);
+	g_assert (name || tb);
+	MonoReflectionAssembly * const ret = name
+		? mono_domain_try_type_resolve_name (domain, name, &error)
+		: mono_domain_try_type_resolve_builder (domain, (MonoReflectionTypeBuilder *)tb, &error);
 	mono_error_cleanup (&error);
-
 	return ret;
 }
 
-MonoReflectionAssembly *
-mono_domain_try_type_resolve_checked (MonoDomain *domain, char *name, MonoObject *tb, MonoError *error)
+/**
+ * mono_class_get_appdomain_do_type_resolve_method:
+ *
+ * This routine returns System.AppDomain.DoTypeResolve.
+ */
+static MonoMethod *
+mono_class_get_appdomain_do_type_resolve_method (void)
 {
 	static MonoMethod *method = NULL;
-	MonoReflectionAssembly *ret;
-	void *params [1];
 
-	error_init (error);
-
-	g_assert (domain != NULL && ((name != NULL) || (tb != NULL)));
+	if (method)
+		return method;
 
+	method = mono_class_get_method_from_name (mono_class_get_appdomain_class (), "DoTypeResolve", -1);
```

Comment:
While you are modifying this code, could you also thread the `MonoError` from `mono_class_get_method_from_name_flags` back to here, so the `g_warning` at https://github.com/mono/mono/pull/6196/files#diff-d51dc4dc4a7d7b173dc78504e0be7fa7R746 can have some more meaningful information? This would entail adding a `mono_class_get_method_from_name_checked` and `mono_class_get_method_from_name_flags_checked` since they are both public API functions.

Reply:
ok. How about we just call it mono_class_get_method? Too short?

Reply:
Eh I see mono_class_get_method_by_index so just plain mono_class_get_method is ambiguous.. these names seem a little bad but when we have so many overloads in C, no great alternative. mono_class_get_method_from_name_flags_checked.

Reply:
How about mono_class_get_method_from_name_checked? i.e. don't state flags?